### PR TITLE
fix(interchain-token-service): prevent large flow_limit from causing failures

### PR DESCRIPTION
**Pull request by TanvirDeol**
[AXE-8254](https://axelarnetwork.atlassian.net/jira/software/c/projects/AXE/boards/21?selectedIssue=AXE-8254)

[AXE-8254]: https://axelarnetwork.atlassian.net/browse/AXE-8254?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: contracts/stellar-interchain-token-service/src/flow_limit.rs [hunk 1]
```diff
@@ -56,18 +56,24 @@ impl FlowDirection {
 
         ensure!(flow_amount <= flow_limit, ContractError::FlowLimitExceeded);
 
-        let new_flow = self
-            .flow(env, token_id.clone())
-            .checked_add(flow_amount)
+        let flow_amount_ = flow_amount as u128;
+        let flow_limit_ = flow_limit as u128;
```
Comment (milapsheth) [inline contracts/stellar-interchain-token-service/src/flow_limit.rs:60]:
```suggestion
        let flow_amount = flow_amount as u128;
        let flow_limit = flow_limit as u128;
```
shadowing the previous names is more idiomatic in rust

File: contracts/stellar-interchain-token-service/src/flow_limit.rs [hunk 2]
```diff
@@ -56,18 +56,24 @@ impl FlowDirection {
...
+        let current_flow_ = self.flow(env, token_id.clone()) as u128;
+        let reverse_flow_ = self.reverse_flow(env, token_id.clone()) as u128;
```
Comment (milapsheth) [inline contracts/stellar-interchain-token-service/src/flow_limit.rs:62]:
```suggestion
        let flow = self.flow(env, token_id.clone()) as u128;
        let reverse_flow = self.reverse_flow(env, token_id.clone()) as u128;
```
avoid `_` suffix

File: contracts/stellar-interchain-token-service/src/tests/flow_limit.rs [hunk 3]
```diff
@@ -353,77 +353,3 @@ fn add_flow_out_fails_on_exceeding_flow_limit() {
         ContractError::FlowLimitExceeded
     );
 }
-
-#[test]
-fn add_flow_fails_on_flow_comparison_overflow() {
```
Comment (milapsheth) [inline contracts/stellar-interchain-token-service/src/tests/flow_limit.rs:358]: old test is being removed, but you are not testing the new behaviour. This needs some tests to ensure overflow is being handled correctly

File: contracts/stellar-interchain-token-service/src/flow_limit.rs [hunk 4]
```diff
@@ -56,18 +56,24 @@ impl FlowDirection {
...
+            .checked_add(flow_limit_)
             .ok_or(ContractError::FlowAmountOverflow)?;
 
         // Equivalent to flow_amount + flow - reverse_flow <= flow_limit
-        ensure!(new_flow <= max_allowed, ContractError::FlowLimitExceeded);
+        ensure!(new_flow_ <= max_allowed_, ContractError::FlowLimitExceeded);
 
+        let new_flow: i128 = new_flow_
+            .try_into()
+            .map_err(|_| ContractError::FlowAmountOverflow)?;
```
Comment (milapsheth) [inline contracts/stellar-interchain-token-service/src/flow_limit.rs:76]: this can still error since `new_flow` could exceed `i128::MAX`, the changes just make the error happen on this line instead of earlier. So, the fundamental logic needs to be changed. And this needs some unit tests

File: contracts/stellar-interchain-token-service/src/flow_limit.rs [hunk 5]
```diff
@@ -56,18 +56,24 @@ impl FlowDirection {
...
-            .checked_add(flow_limit)
+        let max_allowed_ = reverse_flow_
+            .checked_add(flow_limit_)
             .ok_or(ContractError::FlowAmountOverflow)?;
 
         // Equivalent to flow_amount + flow - reverse_flow <= flow_limit
-        ensure!(new_flow <= max_allowed, ContractError::FlowLimitExceeded);
+        ensure!(new_flow_ <= max_allowed_, ContractError::FlowLimitExceeded);
 
+        let new_flow: i128 = new_flow_
```
Comment (milapsheth) [inline contracts/stellar-interchain-token-service/src/flow_limit.rs:74]: to avoid overflow issues, I think the `flow_amount` type needs to be stored as `u128` instead to give more room in the event that `flow_limit` is a large `i128` value

**Review by milapsheth — CHANGES_REQUESTED** [review]

File: contracts/stellar-interchain-token-service/src/migrate.rs [hunk 6]
```diff
@@ -74,7 +74,6 @@ pub fn migrate_token(
         &storage::token_manager_wasm_hash(env),
         &vec![env, ().into()],
     );
-
```
Comment (milapsheth) [inline contracts/stellar-interchain-token-service/src/migrate.rs:77]: why remove this?

File: contracts/stellar-interchain-token-service/src/migrate.rs [hunk 7]
```diff
@@ -93,10 +92,10 @@ pub fn migrate_token(
     };
 
     if let Some(flow_out) = legacy_storage::try_flow_out(env, flow_key.clone()) {
-        storage::set_flow_out(env, token_id.clone(), current_epoch, &flow_out);
+        storage::set_flow_out(env, token_id.clone(), current_epoch, &(flow_out as u128));
```
Comment (milapsheth) [inline contracts/stellar-interchain-token-service/src/migrate.rs:95]:
```suggestion
        storage::set_flow_out(env, token_id.clone(), current_epoch, &u128::try_from(flow_out).expect("expected positive"));
```
avoid implicit casts

File: contracts/stellar-interchain-token-service/src/tests/flow_limit.rs [hunk 8]
```diff
@@ -240,7 +240,7 @@ fn flow_limit_resets_after_epoch() {
 
     execute_its_transfer(&env, &client, &gateway, &token.id, amount);
 
-    assert_eq!(client.flow_in_amount(&token.id), amount);
+    assert_eq!(client.flow_in_amount(&token.id), amount as u128);
```
Comment (milapsheth) [inline contracts/stellar-interchain-token-service/src/tests/flow_limit.rs:243]: make `amount` into a `u128`

File: contracts/stellar-interchain-token-service/src/tests/migrate.rs [hunk 9]
```diff
@@ -311,8 +312,8 @@ fn migrate_native_interchain_token_succeeds() {
         migration_data,
         Some(FlowData {
             token_id,
-            flow_in_amount,
-            flow_out_amount,
+            flow_in_amount: flow_in_amount as u128,
+            flow_out_amount: flow_out_amount as u128,
```
Comment (milapsheth) [inline contracts/stellar-interchain-token-service/src/tests/migrate.rs:316]:
```suggestion
            flow_in_amount,
            flow_out_amount,
```
why cast again?

File: contracts/stellar-interchain-token-service/src/flow_limit.rs [hunk 10]
```diff
@@ -54,10 +54,14 @@ impl FlowDirection {
...
 
-        let new_flow = self
-            .flow(env, token_id.clone())
+        let flow_amount = u128::try_from(flow_amount).expect("expected positive");
+        let flow_limit = u128::try_from(flow_limit).expect("expected positive");
+        let flow = self.flow(env, token_id.clone());
+
+        let new_flow = flow
             .checked_add(flow_amount)
             .ok_or(ContractError::FlowAmountOverflow)?;
```
Comment (ahramy) [inline contracts/stellar-interchain-token-service/src/flow_limit.rs:66]: `FlowAmountOverflow` still exists in the code but has never been covered by a unit test. I think either we should remove the error code since it would not happen, or we should test it assuming it could happen.
